README.md
<p align="center">
  <img title="backbone marionette" src='http://tp4.sinaimg.cn/2179686555/50/5657509044/1' />
</p>

<p align="center">The Django framework</p>

## 概览
**果库网站 V4**

### 系统环境
* Ubuntu 
* Python2.7 
* MySQL 
* Nginx 
* Redis

### 安装 MySQL
```
brew install mysql
```

### 安装 ImageMagick
```
brew install imagemagick
```

### 快速安装
 ```
 virtualenv guoku-v4
 source /path/to/guoku-v4/bin/active
 pip install -r requirements.txt
 ```


### 后端框架
* *django-1.6.5* [相关文档](https://docs.djangoproject.com/en/1.6/)

### 前端框架
* *bootstrap-3.2.0* [相关文档](http://getbootstrap.com/)
* *jQuery-1.11.1*	[相关文档](http://jquery.com/)


<!--### 搜索引擎
* *whoohe-2.5.6* [相关文档](https://pythonhosted.org/Whoosh/index.html)-->


### 安装包
* *django-storage （果库开发）*  [https://github.com/guoku/django-storages.git](https://github.com/guoku/django-storages.git)
* *MySQL-python*
* *django-debug-toolbar*
* *Wand* [相关文档](http://docs.wand-py.org/en/0.3.7/)

### 分词
* "结巴"中文分词 [相关文档](https://github.com/fxsjy/jieba)
```
pip install jieba
```


### Debug 工具
* *django-debug-toolbar* [相关文档](http://django-debug-toolbar.readthedocs.org/en/1.2/installation.html#quick-setup)

### 部署到服务器
* deploy 目录结构

```
deploy/
├── config.ini
├── guoku.pass
├── reload_server.py
├── update_online_code.sh
└── upload_code.py
```

* *config.ini* 配置文件
* *guoku.pass* 密码文件
* *reload_server.py* 重启服务
* *upload_code.py* 上传代码
* *update_online_code.sh* 更新代码 并且重启服务 


```
pip install fabric
cd deploy/
sh update_online_code.sh
```

## Models 与 Cache 之间的关系

＊ 以获取热门商品为例：

利用 django models queryset 实现扩展的 SQL 查询。 Manager 方法来实现缓存。 以减少 models 层面的改动，而影响 views.

```
class EntityLikeQuerySet(models.query.QuerySet):

    def popular(self):
        dt = datetime.now()
        days = timedelta(days=7)
        popular_time = (dt - days).strftime("%Y-%m-%d") + ' 00:00'
        return self.filter(created_time__gt=popular_time).annotate(dcount=models.Count('entity')).values_list('entity_id', flat=True)

    def user_like_list(self, user, entity_list):

        return self.filter(entity_id__in=entity_list, user=user).values_list('entity_id', flat=True)


class EntityLikeManager(models.Manager):

    def get_query_set(self):
        return EntityLikeQuerySet(self.model, using=self._db)

    def popular(self):
        res = cache.get('entity_popular')
        if res:
            return res
        res = self.get_query_set().popular()
        cache.set('entity_popular', res, timeout=3600)
        return res

    def user_like_list(self, user, entity_list):

        return self.get_query_set().user_like_list(user=user, entity_list=entity_list)
```


## 导出数据
```
/opt/mysql5/bin/mysqldump --single-transaction --flush-logs -u root core --ignore-table=core.core_entity_like --ignore-table=core.notifications_notification > core.sql
```

### Email
 <jiaxin@guoku.com>


change_log.md
Change Log
==========
### Ver 4.3.4
1. 分类增加拼音排序
2. 商品搜索 title 增加拼音查找

```
pip install xpinyin
```
```
alter table core_sub_category add column `alias` varchar(128) NOT NULL AFTER `title`;
```

### Ver 4.3.3
1. 购买链接增加防盗链
2. 精选页面 UI 更新
3. 修复添加商品却是店铺信息 bug
4. 后台管理 UI 升级


### 注:
```
ALTER TABLE `core`.`core_buy_link` ADD COLUMN `foreign_price` DECIMAL(20,2) NULL COMMENT '' AFTER `seller`;
```



### Ver 4.3.2
1. 更新用户个人页 ui
2. 商品单页 购买链接 改为服务器跳转。（除淘宝）
3. webUI 更新（专题页）
4. 专题BANNER 增加了辅助字段。

ALTER TABLE `core`.`core_event_banner` 
ADD COLUMN `background_image` VARCHAR(255) NULL DEFAULT NULL AFTER `updated_time`,
ADD COLUMN `background_color` VARCHAR(14) NULL DEFAULT 'fff' AFTER `background_image`;


### Ver 4.3.1
1. v4 增加 discover 
2. solr 替换 sphinx 搜索引擎
3. 增加店铺链接
4. 优化搜索结果

```
ALTER TABLE `core`.`core_buy_link` ADD COLUMN `foreign_price` DECIMAL(20,2) NULL COMMENT '' AFTER `seller`;
```


### Ver 4.3
1. 后台图文编辑增加标签 
2. 修正添加商品 bug
3. category 增加封面图片
4. 更新 404， 500 页面

```
alter table core_category add column `cover` varchar(255) NOT NULL;
```

### Ver 4.2.9
1. 重构标签
2. 页面 title 关键词调整

### Ver 4.2.8
1. 支持 emoj
2. 后台支持 buy link 检索
3. 分类页，去除以下架的商品
4. 图文上线

### Ver 4.2.7（2015-07-06）
1. 增加商品发布到 U站
2. 增加 django-debug-toolbar 
3. 增加 商品喜爱计数器缓存
4. 支持 果库内嵌浏览器 去除导航栏， 测试用 safari ， 改了 user agent  string , 加 orange 进去， 访问 http://test.guoku.com/articles/9/ ， 应该不出现导航栏。  

### Ver 4.2.6（2015-07-02）
1. 完善推荐算法， 去除部分下架商品。
2. 上传 png 自动转换成 jpeg

### Ver 4.2.5 （2015-07-01）
1. spider update 可以检查 淘宝商品价格。 
2. 增加 amazon 图书入库。
3. 部署计数器，和保存阅读数

```
alter table （SQL COMMAND）
ALTER TABLE `core`.`core_article` 
ADD COLUMN `read_count` INT ZEROFILL UNSIGNED NULL AFTER `showcover`;
```

2.  部署代码

2.5 测试计数器工作正常

3. sudo crontab -e   , 打开的编辑器里加入下面一行,并保存

*/30 * * * * /usr/local/bin/python /data/www/nut/script/counter/article_counter_save.py 

### Ver 4.2.4
1. 增加 Article (图文) Sitemap
2. 增加 Article (图文) Rss
3. bug fixes


### Ver 4.2.3 （2015-06-24）


1. **core_buy_link** 表增加 **status** 字段。
2. **core_buy_link** 表增加 **shop_link** 字段。
3. **core_buy_link** 表增加 **seller** 字段。
4. 优化 **nickname** 的显示 
5. 增加微信登陆


### 注：

```
ALTER TABLE core_buy_link ADD COLUMN `status` integer UNSIGNED NOT NULL;
alter table core_buy_link add column `shop_link` varchar(255);
alter table core_buy_link add column `seller` varchar(255);
```





nut/urls.py
from django.conf.urls import patterns, include, url
from django.conf import settings
from django.contrib.staticfiles.urls import staticfiles_urlpatterns
from apps.web.feeds import SelectionFeeds, ArticlesFeeds


urlpatterns = staticfiles_urlpatterns()
# from django.contrib import admin
# admin.autodiscover()


handler500 = 'apps.web.views.page_error'
handler404 = 'apps.web.views.webpage_not_found'
handler403 = 'django.views.defaults.permission_denied'


urlpatterns += patterns(
    '',
    url(r'^403/$', 'django.views.defaults.permission_denied'),
    url(r'^404/$', 'apps.web.views.webpage_not_found'),
    url(r'^500/$', 'apps.web.views.page_error'),
    # (r'^visit_item/$', 'mobile.views.old_visit_item'),
)

urlpatterns += patterns(
    'apps.images',
    url(r'^(images|img|avatar)/', include('apps.images.urls')),
)

urlpatterns += patterns('',

    url(r'^management/', include('apps.management.urls')),
    url(r'^mobile/v3/', include('apps.mobile.urls')),
    url(r'^mobile/v4/', include('apps.v4.urls')),
    url(r'^api/', include('apps.api.urls')),
    url(r'^wechat/', include('apps.wechat.urls')),
    url(r'^counter/', include('apps.counter.urls')),
    url(r'^tag/', include('apps.tag.urls')),

    url(r'^', include('apps.web.urls')),
)



from apps.web.sitemaps import UserSitemap, \
    EntitySitemap, \
    TagSitemap, \
    CategorySitemap, \
    ArticleSitemap

sitemaps = {
    'user': UserSitemap,
    'entity': EntitySitemap,
    'tag': TagSitemap,
    'category': CategorySitemap,
    'article': ArticleSitemap,
}

urlpatterns += patterns(
    'django.contrib.sitemaps.views',
    url(r'^sitemap\.xml$', 'index', {'sitemaps': sitemaps}),
    url(r'^sitemap-(?P<section>.+)\.xml$', 'sitemap', {'sitemaps': sitemaps}),

    url(r'^feed/selection/$', SelectionFeeds()),
    url(r'^feed/articles/$', ArticlesFeeds()),
)

if settings.DEBUG:
    import debug_toolbar
    urlpatterns += patterns('',
        url(r'^__debug__/', include(debug_toolbar.urls)),
    )